File: backend/app/documents.py
"""Helpers and input models for production documents."""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field

from .models import ClearanceItem, ProductionDocument
from .scope import DocumentScope, IntendedUseProfile, ScopeAssessment, assess_scope


def comma_list(value: str) -> list[str]:
    return list(dict.fromkeys(part.strip() for part in value.split(",") if part.strip()))


def find_document(item: ClearanceItem, document_id: str) -> Optional[ProductionDocument]:
    return next((document for document in item.documents if document.id == document_id), None)


def document_scope(document: ProductionDocument) -> DocumentScope:
    return DocumentScope(
        media=document.media,
        territories=document.territories,
        starts_on=document.starts_on,
        ends_on=document.ends_on,
        perpetual=document.perpetual,
        covered_use=document.covered_use,
    )


def scope_for(document: ProductionDocument, intended: IntendedUseProfile) -> ScopeAssessment:
    return assess_scope(intended, document_scope(document))


class DocumentMetadataPatch(BaseModel):
    kind: Optional[str] = None
    title: Optional[str] = Field(default=None, max_length=180)
    notes: Optional[str] = Field(default=None, max_length=4000)
    media: Optional[list[str]] = None
    territories: Optional[list[str]] = None
    starts_on: Optional[str] = None
    ends_on: Optional[str] = None
    perpetual: Optional[bool] = None
    covered_use: Optional[str] = Field(default=None, max_length=1000)

File: apps/api/src/services/documents.ts
import type {
  DocumentMetadataPatch,
  ProductionDocument,
  Project,
  ScopeAssessment,
} from "@clearcut/contracts";
import { assessScope } from "@clearcut/domain";

import { ApiProblem } from "../middleware/errors";
import { audit, requireItem } from "./items";

export function commaList(value: string): string[] {
  return value.split(",").map((part) => part.trim()).filter(Boolean);
}

export function documentScope(project: Project, document: ProductionDocument): ScopeAssessment {
  return assessScope(project.use_profile, {
    media: document.media,
    territories: document.territories,
    starts_on: document.starts_on,
    ends_on: document.ends_on,
    perpetual: document.perpetual,
    covered_use: document.covered_use,
  });
}

export function requireDocument(project: Project, itemId: string, documentId: string): ProductionDocument {
  const item = requireItem(project, itemId);
  const document = item.documents.find((candidate) => candidate.id === documentId);
  if (!document) throw new ApiProblem(404, "document_not_found", "Document not found for this item.");
  return document;
}

export function patchDocument(
  project: Project,
  itemId: string,
  documentId: string,
  patch: DocumentMetadataPatch,
): ProductionDocument {
  const item = requireItem(project, itemId);
  const document = requireDocument(project, itemId, documentId);
  Object.assign(document, patch);
  item.audit_events.push(audit({
    actor: "coordinator",
    action: "document_metadata_updated",
    rationale: `Updated recorded scope for ${document.title}.`,
    sourceVersion: item.source_version,
    detail: { document_id: document.id, fields: Object.keys(patch).sort() },
  }));
  return document;
}
